Tests/AutoCollection/Heartbeat.tests.ts

```diff
@@ -0,0 +1,249 @@
+                assert.equal(properties2["sdk"], Context.sdkVersion, "sdk version is read from Context");
+                assert.equal(properties2["osType"], os.type(), "osType is read from os library");
+                assert.equal(properties2["azfunction_appId"], "host_name", "azfunction_appId is read from envrionment variable");
+
+                process.env = originalEnv;
+                stub2.restore();
+                heartbeat2.dispose();
+                done();
+            });
+        });
+
+        it("should read correct VM information from response", (done) => {
+            const heartbeat3: HeartBeat = new HeartBeat(client);
+            heartbeat3.enable(true, client.config);
+            HeartBeat.INSTANCE.enable(true, client.config);
+            const stub3 = sinon.stub(heartbeat3["_client"], "trackMetric");
+            const requestURL = "http://169.254.169.254"
+            nock(requestURL, {
+                reqheaders: {
+                    "Metadata": "true"
+                }
+            })
+            .get('/metadata/instance/compute?api-version=2017-12-01&format=json')
+            .reply(
+                200,
+                {
+                    'vmId': 1,
+                    'subscriptionId': 2,
+                    'osType': 'Linux'
+                }
+            );
+
+            heartbeat3["trackHeartBeat"](client.config, () => {
+                assert.equal(stub3.callCount, 1, "calls trackMetric for the VM heartbeat metric");
+                assert.equal(stub3.args[0][0].name, "HeartBeat", "uses correct name for heartbeat metric");
+                assert.equal(stub3.args[0][0].value, 0, "checks value is 0");
+                const keys3 = Object.keys(stub3.args[0][0].properties);
+                assert.equal(keys3.length, 4, "makes sure 4 kv pairs are added when resource type is VM");
+                assert.equal(keys3[0], "sdk", "sdk is added as a key");
+                assert.equal(keys3[1], "azInst_vmId",  "azInst_vmId is added as a key");
+                assert.equal(keys3[2], "azInst_subscriptionId", "azInst_subscriptionId is added as a key");
+                assert.equal(keys3[3], "azInst_osType", "azInst_osType is added as a key");
+
+                const properties3 = stub3.args[0][0].properties;
+                assert.equal(properties3["sdk"], Context.sdkVersion, "sdk version is read from Context");
+                assert.equal(properties3["azInst_vmId"], 1, "azInst_vmId is read from response");
+                assert.equal(properties3["azInst_subscriptionId"], 2, "azInst_subscriptionId is read from response");
+                assert.equal(properties3["azInst_osType"], "Linux", "azInst_osType is read from response");
+
+                stub3.restore();
+                heartbeat3.dispose();
+                done();
+            });
+        });
+
+        it("should read correct VM information from response", (done) => {
```

Comment:
@markwolff @MSNev I'm still not able to make this VM test work.. I have two versions. heartbeat3 is using nock... heartbeat4 is to stub out makeRequest, but I'm not sure how to give the callback of "makeRequest" a parameter (fake res)? please see comment below.

Reply:
As your using sinon, you could use the sinon.useFakeServer() which provides a mock XHR request that you can respond() with the status and values inline -- rather than wait for a real response etc